proposed/2022/NuGetExeWarnWhenSDKProjectsGetPacked.md

```diff
@@ -0,0 +1,99 @@
+```
+
+This *breaking change* will be announced via blog well before the NuGet.exe carrying the change ships.
+
+### Technical explanation
+
+As described above, whenever there's no SDK installed, there's simply no accurate way for NuGet to detect that a project is SDK-based.
+We can do some XML parsing, but it is likely to be error prone. `NuGet.exe pack` is not the only command that will fail, so the customer will likely figure out that they need to install the .NET SDK. After that the same scenarios with the installed .NET SDK will follow.
+
+To determine whether a project is `.NET SDK` based, we will evaluate the project (part of the normal pack operation), and check for the `UsingMicrosoftNETSDK` property being set. If the property is set, then the project is presumed to be .NET SDK based and will be blocked from packing.
+
+## Drawbacks
+
+- The proposal here would be `breaking` for customers, but it's more than likely that the scenario for which they used nuget.exe pack on an SDK based project never worked correctly.
+
+## Rationale and alternatives
+
+- Do nothing. This would mean just acknowledging that SDK-based projects support in NuGet.exe would not be added and not do anything to help migrate users to `dotnet.exe pack` and `msbuild /t:pack`.
+- Add full pack SDK support into NuGet.exe. While not impossible, this creates maintainability considerations. dotnet.exe has a more involved support for the SDK-based PackageReference projects, and as such is considered to be the `right tool for the job`.
+- Call `msbuild /t:pack` from NuGet.exe. This has the same considerations as the previos approach that `dotnet.exe` should be recommended as the right tool for the job.
+This is also likely to quietly change the behavior for certain users and change the type of package they're generating.
+A change like this likely has a lot more merit in 2018, than 2022 given how long dotnet.exe has been the recommended tool for SDK based projects..
+- Add a warning and proceed packing.
+  - A behavior change to an error in a minor version can be considered disruptive. This is one of those special cases where the disruption might be warranted.
+  Instead of an error, we could add a warning instead. Given that NuGet.exe pack by default will usually raise a bunch of warnings, it is likely that if warnings were good enough, the customers would have reviewed said warnings and figured out that all those warnings cannot be addressed correctly.
+- Add an information message and proceed warning. Given that the information this tool is not the right one for the job is critical, this is not likely to move the needle enough.
+
+## Prior Art
+
+N/A
+
+## Unresolved Questions
+
+## Future Possibilities
+
+- We can consider *warning* or *erroring* when `NuGet.exe pack` is being used on csproj PackageReference projects. Given that you need to undergo a manual migration to PackageReference, this isn't likely to be a priority scenario.

```

Comment:
Since `nuget pack` does not `PackageReference` why not error out with the escape hatch similar to this proposal and make `nuget pack` call Pack target for both PC and SDK-style projects if pack targets are imported in both of them.

I think this would be a better solution for customers. This solves two scenarios in one go. It'll also help people migrating to .NET (Core).

We could add a suggestion to use `dotnet pack` as a warning later on if MSBuild tries to break compat. It will have the least impact as it would be tie in with MSBuild or Visual Studio announcement.

Reply:
I think it'd be trivial to handle the old csproj PackageReference scenario once we have established a pattern here. 
By keeping it scoped, it increases the chance of this getting accepted. 

> We could add a suggestion to use dotnet pack as a warning later on if MSBuild tries to break compat. It will have the least impact as it would be tie in with MSBuild or Visual Studio announcement.

There's really no difference between `msbuild /t:pack` and `dotnet pack`. 
The difference is really between old csproj and SDK csproj, as the SDK one automatically includes the pack targets, while the old style does not, thus it does not have the same servicing as SDK projects.

Reply:
I don't think my message is getting through. What I mean is that there's a chance of getting this proposal to work for both PR and SDK-style by checking for NuGet Pack targets import. There's property `ImportNuGetBuildTasksPackTargetsFromSdk` but that doesn't really tell us if the user has disabled Pack targets or if the targets has been imported.

We could add a new property `NuGetPackTargetsHasBeenImported` to Pack targets and check this for Pack target availability. If the Pack targets are available then we would invoke `msbuild -t:Pack` and call it a day. Bonus points for checking if the project is using `packages.config` or `PackageReference` restore style.

If the property is not available (_it could be older targets_) then there are 3 ways to deal with them...

1. Check for `ImportNuGetBuildTasksPackTargetsFromSdk` property or `NuGet.Build.Tasks.Pack` package. If it is true then proceed to `msbuild -t:Pack` else continue.

2. If the project has set `UsingMicrosoftNETSdk` but disabled `UsingNETSdkDefaults`, it is likely a legacy project using .NET SDK (_yeah, seen those lying around in the GitHub_). If that is the case then show the suggestion to enable Pack targets and error out with an escape hatch.

3. Finally, if the project is legacy and using `packages.config` then proceed to `nuget pack`. If it is using `ProjectReference` then show suggestion to enable Pack targets and error out with an escape hatch.

The escape hatch should always lead to `nuget pack`. This is just a starting point but things could change during actual implementation.

Reply:
I don't think my message is getting through. What I mean is that there's a chance of getting this proposal to work for both PR and SDK-style by checking for NuGet Pack targets import. There's property `ImportNuGetBuildTasksPackTargetsFromSdk` but that doesn't really tell us if the user has disabled Pack targets or if the targets has been imported.

We could add a new property `NuGetPackTargetsHasBeenImported` to Pack targets and check this for Pack target availability. If the Pack targets are available then we would invoke `msbuild -t:Pack` and call it a day. Bonus points for checking if the project is using `packages.config` or `PackageReference` restore style.

If the property is not available (_it could be older targets_) then there are 3 ways to deal with them...

1. Check for `ImportNuGetBuildTasksPackTargetsFromSdk` property or `NuGet.Build.Tasks.Pack` package. If it is true then proceed to `msbuild -t:Pack` else continue.

2. If the project has set `UsingMicrosoftNETSdk` but disabled `UsingNETSdkDefaults`, it is likely a legacy project using .NET SDK (_yeah, seen those lying around in the GitHub_). If that is the case then show the suggestion to enable Pack targets and error out with an escape hatch.

3. Finally, if the project is legacy and using `packages.config` then proceed to `nuget pack`. If it is using `ProjectReference` then show suggestion to enable Pack targets and error out with an escape hatch.

The escape hatch should always lead to `nuget pack`. This is just a starting point but things could change during actual implementation.

Reply:
@Nirmal4G 
I get your point, we're just explicitly choosing *not* to do it in this change.  That's what it's listed in future possibilities. 

> There's property ImportNuGetBuildTasksPackTargetsFromSdk but that doesn't really tell us if the user has disabled Pack targets or if the targets has been imported.

Yeah, there's no such property, but there's a capability that we might be able to use instead.

Reply:
See NuGet/NuGet.Client#1368. This PR add the `$(ImportNuGetBuildTasksPackTargetsFromSdk)` property. And it's been there ever since.

Reply:
`NugetTaskAssemblyFile` is an even older one. We have options for sure.

Reply:
Resolving this thread.